server/service/src/invoice/outbound_return/generate_outbound_return_lines.rs

```diff
@@ -0,0 +1,148 @@
+use crate::{service_provider::ServiceContext, ListError, ListResult};
+use repository::{EqualFilter, StockLine, StockLineFilter, StockLineRepository};
+use util::uuid::uuid;
+
+#[derive(Debug, Clone)]
+pub struct OutboundReturnLine {
+    pub id: String,
+    pub reason_id: Option<String>,
+    pub comment: Option<String>,
+    pub number_of_packs: u32,
+    pub stock_line: StockLine,
+}
+
+pub fn generate_outbound_return_lines(
+    ctx: &ServiceContext,
+    store_id: &str,
+    stock_line_ids: Vec<String>,
+    item_id: Option<String>,
+    return_id: Option<String>,
+) -> Result<ListResult<OutboundReturnLine>, ListError> {
+    if stock_line_ids.is_empty() && item_id.is_none() && return_id.is_none() {
+        return Ok(ListResult {
+            count: 0,
+            rows: vec![],
+        });
+    }
+
+    let mut filter = StockLineFilter::new().is_available(true);
+
+    if !stock_line_ids.is_empty() {
+        filter.id = Some(EqualFilter::equal_any(
+            stock_line_ids.iter().map(String::clone).collect(),
+        ))
+    }
+    filter.item_id = item_id.map(|item_id| EqualFilter::equal_to(&item_id));
```

Comment:
With setting the filter up like this, if you pass in both stock line ids and an item id, we would only get back stock lines of that item... is that what you would expect of this API?

In practice, I think we'll only use one or the other - generate from stock_line_ids in inbound shipment, or generate from an item_id in the return detail view.

Other options:
- Throw an error - ask them to choose one or the other
- Make it an OR filter rather than AND

Reply:
I wish we could do union inputs in grahpql (from memory I think async-grahpql actually added it, but it's not part of the standard graphql).

I think what you've done is fine, maybe a comment in graphql about this (btw async-graphql should expose comments [like this](https://github.com/msupply-foundation/open-msupply/blob/ae7135cb1ae2b6cd08f0ef3f92d558c04e021c90/server/graphql/types/src/types/invoice_query.rs#L38-L42) as introspectable comments)

Reply:
Hmm this is a bit trickier on second thought, because we need to only show available stock_lines when item is selected (don't want to list all of the stock_lines that ever existing for that item), but when stock_lines are selected we need to show even if not available (so that user doesn't question, 'why' i couldn't return the thing i selected). What makes it ever more spicy is that if item is selected and return_id is selected, if stock_lines is not available, but we already return some quantity for that stock_line in return, it should be listed. 
I think this would be a great unit test use case, and I think the logic would mean, get stock_lines selected in stock_line_ids, get stock_lines and invoice_lines from return_id (for that item), get all stock_lines for that item that are available, then consolidate stock_lines that we've just selected and add numberOfPacks, reason, comment from invoice_lines

Reply:
ooo that is spicy! yes definitely will want some tests around that one 😅 Will be fun to sink teeth into soon!